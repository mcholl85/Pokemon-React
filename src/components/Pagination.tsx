import { PAGINATION_DELTA } from '../constants'
import { getRangeOfPages } from '../utils/getRangePages'

type PaginationProps = {
  currentPage: number
  totalPage?: number
  setPage: (params: { [key: string]: string }) => void
}

function Pagination({ currentPage, totalPage = 1, setPage }: PaginationProps) {
  const range = getRangeOfPages(currentPage, totalPage, PAGINATION_DELTA)
  const handleClick = (page: number) => setPage({ page: String(page) })

  return (
    <nav aria-label='navigation' className='mb-2'>
      <ul className='pagination flex-wrap justify-content-start'>
        <li className='page-item px-1'>
          <button
            className={`page-link border-0 ${
              currentPage <= 1 ? 'disabled' : 'text-identity rounded-2 shadow-none'
            }`}
            tabIndex={-1}
            onClick={() => handleClick(currentPage - 1)}
          >
            Previous
          </button>
        </li>
        {range.map((page, index) =>
          typeof page === 'number' ? (
            <li key={index} className='page-item px-1'>
              <button
                className={`page-link text-identity border-0 rounded-2 shadow-none ${
                  page === currentPage ? 'active' : ''
                }`}
                onClick={() => handleClick(page)}
              >
                {page}
              </button>
            </li>
          ) : (
            <li key={index} className='page-item px-1'>
              <button className='page-link text-identity border-0 rounded-2 disabled shadow-none'>
                {page}
              </button>
            </li>
          ),
        )}
        <li className='page-item px-1'>
          <button
            className={`page-link border-0 ${
              currentPage >= totalPage ? 'disabled' : 'text-identity rounded-2 shadow-none'
            }`}
            onClick={() => handleClick(currentPage + 1)}
          >
            Next
          </button>
        </li>
      </ul>
    </nav>
  )
}
export default Pagination
