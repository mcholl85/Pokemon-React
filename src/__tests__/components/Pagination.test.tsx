import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Pagination from '../../components/Pagination'

describe('Pagination', () => {
  test('Matches DOM snapshot', () => {
    const { asFragment } = render(<Pagination currentPage={2} totalPage={2} setPage={() => ({})} />)

    expect(asFragment()).toMatchSnapshot()
  })
  test.each`
    currentPage
    ${1}
    ${2}
    ${3}
  `('Should have the button $currentPage  active', ({ currentPage }) => {
    render(<Pagination currentPage={currentPage} totalPage={currentPage} setPage={() => ({})} />)
    const button = screen.getByRole('button', { name: currentPage })

    expect(button).toHaveClass('active')
  })
  test('Should have the previous button disabled given a totalPage of 3 when the currentPage is 1', () => {
    render(<Pagination currentPage={1} totalPage={3} setPage={() => ({})} />)
    const previousButton = screen.getByRole('button', { name: 'Previous' })
    const nextButton = screen.getByRole('button', { name: 'Next' })
    expect

    expect(previousButton).toHaveClass('disabled')
    expect(nextButton).not.toHaveClass('disabled')
  })
  test('Should have the next button disabled given a totalPage of 3 when the currentPage is 3', () => {
    render(<Pagination currentPage={3} totalPage={3} setPage={() => ({})} />)
    const previousButton = screen.getByRole('button', { name: 'Previous' })
    const nextButton = screen.getByRole('button', { name: 'Next' })
    expect

    expect(previousButton).not.toHaveClass('disabled')
    expect(nextButton).toHaveClass('disabled')
  })
  test('Should set a new page given a click on active button', async () => {
    const setPageMock = jest.fn()
    render(<Pagination currentPage={3} totalPage={3} setPage={setPageMock} />)
    const button = screen.getByRole('button', { name: '2' })

    await userEvent.click(button)

    expect(setPageMock).toBeCalledTimes(1)
    expect(setPageMock).toBeCalledWith({ page: '2' })
  })
})
