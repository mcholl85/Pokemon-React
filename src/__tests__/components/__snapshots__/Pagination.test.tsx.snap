// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Matches DOM snapshot 1`] = `
<DocumentFragment>
  <nav
    aria-label="navigation"
    class="mb-2"
  >
    <ul
      class="pagination flex-wrap justify-content-start"
    >
      <li
        class="page-item px-1"
      >
        <button
          class="page-link border-0 text-identity rounded-2 shadow-none"
          tabindex="-1"
        >
          Previous
        </button>
      </li>
      <li
        class="page-item px-1"
      >
        <button
          class="page-link text-identity border-0 rounded-2 shadow-none "
        >
          1
        </button>
      </li>
      <li
        class="page-item px-1"
      >
        <button
          class="page-link text-identity border-0 rounded-2 shadow-none active"
        >
          2
        </button>
      </li>
      <li
        class="page-item px-1"
      >
        <button
          class="page-link border-0 disabled"
        >
          Next
        </button>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
